### PR TITLE
feat: add abbreviation.cpp: topic: Dynamic Programming

### PR DESCRIPTION
**Pull request by AshishYUO**
#### Description of Change

New Feature: Abbreviation: A Dynamic Programming problem, to check if converting of a given string is possible [More Info Here](https://www.hackerrank.com/challenges/abbr/problem)
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: dynamic_programming/abbreviation.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+ * 3. If the a[i] is not uppercase, just discard the char, else return false
+ *
+ * Time Complexity (O(|a|*|b|)) where |a| => length of string
+ * @author Ashish Daulatabad (https://github.com/AshishYUO)
+ */
+
+#include <cassert>
+#include <iostream>
+#include <string>
+#include <vector>
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:28]: Add a brief description of what the header/library is for (see the example below).
```c++
#include <cassert>     /// for assert
#include <iostream>    /// for IO operations
```

File: dynamic_programming/abbreviation.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming Algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @namespace Knapsack
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:35]:
```suggestion
 * @namespace abbreviation
```

File: dynamic_programming/abbreviation.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+ * @brief Implementation of Abbreivation problem
+ */
+namespace abbreviation {
+/**
+ *  (recursive dp function) https://www.hackerrank.com/challenges/abbr/problem
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:40]:
```suggestion
 * @brief Recursive DP function
```

File: dynamic_programming/abbreviation.cpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+ *  Returns whether s can be converted to t with following rules:
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:41]:
```suggestion
 *
 * @details
 * Returns whether s can be converted to t with the following rules:
```

File: dynamic_programming/abbreviation.cpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+ *  a. Capitalize zero or more of a's lowercase letters from string s
+ *  b. remove all other lowercase letters from string s
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:43]:
```suggestion
 * a. Capitalize zero or more of a's lowercase letters from string s
 * b. remove all other lowercase letters from string s
```

File: dynamic_programming/abbreviation.cpp [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+ * @param s: given string
+ * @param t: resultant abbreivated string
+ * @param i: start of string s
+ * @param j: start of string j
+ * @returns bool whether s can be converted to t
+ */
+bool abbreviation_recursion(std::vector<std::vector<bool>> *memo,
+                            std::vector<std::vector<bool>> *visited,
+                            const std::string &s, const std::string &t,
+                            int i = 0, int j = 0) {
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:55]: I don't think these should be parameters. They should be variables inside the function.

File: dynamic_programming/abbreviation.cpp [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    }
+    (*memo)[i][j] = ans;
+    (*visited)[i][j] = true;
+    return (*memo)[i][j];
+}
+/**
+ *  (iterative dp function) https://www.hackerrank.com/challenges/abbr/problem:
+ *  Returns whether s can be converted to t with following rules:
+ *  a. Capitalize zero or more of a's lowercase letters from string s
+ *  b. remove all other lowercase letters from string s
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:96]: No need to add a link here. Consider adding a proper brief description of the function.
```suggestion
 * @brief Iterative DP function
 * @details
 * Returns whether s can be converted to t with the following rules:
 * a. Capitalize zero or more of `a`'s lowercase letters from string `s`
 * b. remove all other lowercase letters from string `s`
```

File: dynamic_programming/abbreviation.cpp [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+ * @param s: string
+ * @param t: resultant string
+ * @returns boolean (true or false) whether s can be converted to t
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:99]: These parameters need more proper information.
```suggestion
 * @param s string
 * @param t resultant string
 * @returns boolean (true or false) whether `s` can be converted to `t`
```

File: dynamic_programming/abbreviation.cpp [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+                }
+            }
+        }
+    }
+    return memo.back().back();
+}
+}  // namespace abbreviation
+}  // namespace dynamic_programming
+
+static void test() {
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:126]:
```suggestion
/**
 * @brief Self test-implementations
 * @returns void
 */
static void test() {
```

File: dynamic_programming/abbreviation.cpp [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+    visited = std::vector<std::vector<bool>>(
+        s.size() + 1, std::vector<bool>(t.size() + 1, false));
+
+    assert(dynamic_programming::abbreviation::abbreviation_recursion(
+               &memo, &visited, s, t) == true);
+    assert(dynamic_programming::abbreviation::abbreviation(s, t) == true);
+}
+
+int main() {
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:161]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
```

File: dynamic_programming/abbreviation.cpp [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    test();
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:162]:
```suggestion
    test();  // run self-test implementations
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good code.

File: dynamic_programming/abbreviation.cpp (hunk 6 above)
Comment (AshishYUO) [reply dynamic_programming/abbreviation.cpp:55]: I don't know how to do that actually.

If the intention is to not let user pass memo and visited and let function handle it, then I can create a helper function for that.

Comment (Panquesito7) [reply dynamic_programming/abbreviation.cpp:55]: Just create two variables inside the function.
```suggestion
) {
    int i = 0, j = 0;
```

Comment (AshishYUO) [reply dynamic_programming/abbreviation.cpp:55]: These two variables are helpful for resolving the base case or changing the state from one to another.
E.g., lines 74, 76, and 77.

I kept them default to zero if the user does not pass them explicitly.

Comment (Panquesito7) [reply dynamic_programming/abbreviation.cpp:55]: Yes, that's good, and it looks fine for me. But why not use variables instead of parameters?

Comment (AshishYUO) [reply dynamic_programming/abbreviation.cpp:55]: I don't know if there are other ways around to call recursion without `i` and `j` as parameters.
I can only think of them by passing them as parameters.

File: dynamic_programming/abbreviation.cpp [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+ * Returns whether `s` can be converted to `t` with following rules:
+ * a. Capitalize zero or more of a's lowercase letters from string `s`
+ * b. remove all other lowercase letters from string `s`
+ * @param memo memo as parameter to store the result
+ * @param visited visited boolean to check if the result is already computed
+ * @param str given string, which might not be abbreivated
+ * @param result resultant abbreivated string
+ * @param str_idx pointer for string `str`, helpful for transitions
+ * @param result_idx pointer for string `result`, helpful for transitions
+ * @returns boolean (`true` or `false`) whether string `str` can be converted to
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:52]: Complete the steps below. Fill in the documentation below.
```suggestion
 * @returns `false` if ....
 * @returns `true` if ....
```

File: dynamic_programming/abbreviation.cpp [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+ * 2. Delete all of the remaining lowercase letters in a.
+ *
+ * ##Algorithm
+ * The idea is in the problem statement itself: iterate through characters of
+ * string a and b (for character indexes i and j respectively):
+ * 1. If a[i] and b[j] are equal, then move to next position
+ * 2. If a[i] is lowercase of b[j], then explore two possibilities:
+ * a. Capitalize or
+ * b. Skip a[i]
+ * 3. If the a[i] is not uppercase, just discard the char, else return false
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:19]:
```suggestion
 * The idea is in the problem statement itself: iterate through characters of
 * string `a` and `b` (for character indexes i and j respectively):
 * 1. If `a[i]` and `b[j]` are equal, then move to next position
 * 2. If `a[i]` is lowercase of `b[j]`, then explore two possibilities:
 * a. Capitalize
 * b. Skip `a[i]`
 * 3. If `a[i]` is not uppercase, just discard the char, else return false
```

File: dynamic_programming/abbreviation.cpp [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+ * a. Capitalize `a[i]` or
+ * b. Skip `a[i]`
+ * 3. If the `a[i]` is not uppercase, just discard that character, else return
+ * `false`
+ *
+ * Time Complexity: (O(|a|*|b|)) where `|a|` => length of string `a`
+ * @author [Ashish Daulatabad](https://github.com/AshishYUO)
+ */
+
+#include <cassert>   /// `assert` function
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:26]:
```suggestion
#include <cassert>   /// for `assert`
```

File: dynamic_programming/abbreviation.cpp [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+#include <iostream>  /// input/output operations
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:27]:
```suggestion
#include <iostream>  /// for IO operations
```

File: dynamic_programming/abbreviation.cpp [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+#include <string>    /// `std::string` library
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:28]:
```suggestion
#include <string>    /// for `std::string` library
```

File: dynamic_programming/abbreviation.cpp [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+#include <vector>    /// `std::vector` STL library
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:29]:
```suggestion
#include <vector>    /// for `std::vector` STL library
```

File: dynamic_programming/abbreviation.cpp [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+ * @param result resultant abbreivated string
+ * @param str_idx pointer for string `str`, helpful for transitions
+ * @param result_idx pointer for string `result`, helpful for transitions
+ * @returns `false` if string `str` cannot be converted to `result`
+ * @returns `true` if string `str` can be converted to `result`
+ */
+bool abbreviation_recursion(std::vector<std::vector<bool>> *memo,
+                            std::vector<std::vector<bool>> *visited,
+                            const std::string &str, const std::string &result,
+                            int str_idx = 0, int result_idx = 0) {
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:59]: Consider using `uint64_t`, `uint32_t`, `uint16_t` or even `uint8_t` for non-negative values. For negative values, use `int64_t` or their appropriate size. Check other parts of the code too ([reference](https://en.cppreference.com/w/cpp/types/integer)). 🙂

File: dynamic_programming/abbreviation.cpp [hunk 19]
```diff
@@ -0,0 +1,192 @@
+/**
+ * @file
+ * @brief Implementation of Abbrievation
+ * (https://www.hackerrank.com/challenges/abbr/problem)
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:4]: We use [Markdown](https://guides.github.com/features/mastering-markdown/) format here.
```suggestion
 * @brief [Abbreviation](https://www.hackerrank.com/challenges/abbr/problem) implementation
```

File: dynamic_programming/abbreviation.cpp [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+#include <string>    /// for `std::string` library
+#include <vector>    /// for `std::vector` STL library
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming Algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @namespace abbreviation
+ * @brief Implementation of Abbreivation problem
```
Comment (Panquesito7) [inline dynamic_programming/abbreviation.cpp:37]:
```suggestion
 * @brief Functions for [Abbreviation](https://www.hackerrank.com/challenges/abbr/problem) implementation
```

**Review by Panquesito7 — DISMISSED** [review]
Amazing work, I like how you've improved & made the code so well. 🎉
Code and documentation look good and very refined; LGTM. 😄 👍

**Comment by AshishYUO** [comment]
Thank you for helping me documenting the code. I'm not used to this much, so it took some time for me.

**Comment by AshishYUO** [comment]
My bad. I accidentally clicked on close with comments.

**Comment by Panquesito7** [comment]
> accidentally clicked on close with comments

No worries, it's alright. 😄

File: dynamic_programming/abbreviation.cpp [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+ */
+namespace abbreviation {
+/**
+ * @brief
+ * Recursive Dynamic Programming function
+ * @details
+ * Returns whether `s` can be converted to `t` with following rules:
+ * a. Capitalize zero or more of a's lowercase letters from string `s`
+ * b. remove all other lowercase letters from string `s`
+ * @param memo memo as parameter to store the result
```
Comment (ayaankhan98) [inline dynamic_programming/abbreviation.cpp:49]:
```suggestion
 * @param memo To store the result
```

File: dynamic_programming/abbreviation.cpp [hunk 22]
```diff
@@ -0,0 +1,194 @@
...
+ * @param visited visited boolean to check if the result is already computed
```
Comment (ayaankhan98) [inline dynamic_programming/abbreviation.cpp:50]:
```suggestion
 * @param visited boolean to check if the result is already computed
```

**Review by Panquesito7 — APPROVED** [review]

**Review by ayaankhan98 — APPROVED** [review]
LGTM, @AshishYUO Thanks for your contribution.

**Pull request merged**